### PR TITLE
Add Spark Mapper

### PR DESCRIPTION
**Pull request by mik-laj**
Hello,

I implemented Spark Mapper.

The current implementation has some limitations:
* The following elements are not supported:`job-tracker`, `name-node`, `master`, `mode`.
* The `spark-opts` element is partially converted. Only `conf` option is converted and not all format of value is supported.

Reference:
https://oozie.apache.org/docs/5.1.0/DG_SparkActionExtension.html

Thanks,

### CODE REVIEW

File: oozie-to-airflow/mappers/spark_mapper.py [hunk 1]
```diff
@@ -45,190 +59,171 @@ def __init__(
...
-        # master url, deploy mode,
-        self.application = self.test_and_set(oozie_node, "jar", "''", params=self.params, quote=True)
-        self.spark_name = self.test_and_set(
-            oozie_node, "name", "'airflow-spark'", params=self.params, quote=True
-        )
-        self.java_class = self.test_and_set(oozie_node, "class", None, params=self.params, quote=True)
+        self.java_jar = self._get_or_default(self.oozie_node, SPARK_TAG_JAR, None, params=self.params)
+        self.java_class = self._get_or_default(self.oozie_node, SPARK_TAG_CLASS, None, params=self.params)
+        if self.java_class and self.java_jar:
+            self.dataproc_jars = [self.java_jar]
```
Comment (sprzedwojski) [inline oozie-to-airflow/mappers/spark_mapper.py]: It's worth noting that according to the docs:
> The jar element indicates a comma separated list of jars or python files.

So I see 2 problems here:
1. In this case this may be a one-element list, with the element being a comma-separated list in a string ;)
2. These can be Python files, which might be rejected by the Dataproc API
  - With regards to 2) we could probably write in the docs that we don't support Python files here.

File: oozie-to-airflow/templates/action.tpl [hunk 2]
```diff
@@ -13,8 +13,8 @@
   See the License for the specific language governing permissions and
   limitations under the License.
  #}
-{% for op in sub_ops %}
-{{ op.rendered_template }}
+{% for task in tasks %}
+{{ task.rendered_template }}
```
Comment (sprzedwojski) [inline oozie-to-airflow/templates/action.tpl]: Why are we renaming this template? Even if it's needed, I'd suggest a separate commit/PR.

File: oozie-to-airflow/templates/mapreduce.tpl [hunk 3]
```diff
@@ -24,10 +24,10 @@
     main_class=PARAMS['hadoop_main_class'],
     arguments=['{{ properties['mapreduce.input.fileinputformat.inputdir'] }}', '{{ properties['mapreduce.output.fileoutputformat.outputdir'] }}'],
     {% if hdfs_files %}
-    files=["{{ hdfs_files }}"],
+    files={{ hdfs_files | tojson }},
     {% endif %}
     {% if hdfs_archives %}
-    archives=["{{ hdfs_archives }}"],
+    archives={{ hdfs_archives | tojson }},
```
Comment (sprzedwojski) [inline oozie-to-airflow/templates/mapreduce.tpl]: These changes concern MapReduce, let's do them in a separate PR.

File: oozie-to-airflow/mappers/spark_mapper.py [hunk 4]
```diff
@@ -45,190 +59,171 @@ def __init__(
...
+                trigger_rule=self.trigger_rule,
+            ),
+        )
+
+        if not self.has_prepare(self.oozie_node):
+            return [action_task]
+
+        prepare_task = Task(
+            task_id=self.name + "_prepare",
+            template_name="prepare.tpl",
```
Comment (sprzedwojski) [inline oozie-to-airflow/mappers/spark_mapper.py]: In other mappers (e.g. Pig, Mapreduce) we include the prepare task in the main task's template.
If we want to do this differently maybe let's change it everywhere in a separate commit? It can even be in this PR I think, but what's important is not to have 2 ways of doing "prepare".

File: oozie-to-airflow/mappers/fs_mapper.py [hunk 5]
```diff
@@ -128,18 +128,18 @@ def parse_tasks(self):
...
+                    template_params=dict(trigger_rule=self.trigger_rule),
                 )
             ]
 
         return [self.parse_fs_action(i, node) for i, node in enumerate(self.oozie_node)]
 
     def convert_to_text(self):
         return render_template(
-            template_name="fs.tpl",
+            template_name="action.tpl",
```
[the remark this answers is not in the document]
Comment (mik-laj) [reply oozie-to-airflow/mappers/fs_mapper.py]: Because I changed the name of the template. This template is so universal that it can be used in several places.

File: oozie-to-airflow/mappers/pig_mapper.py [hunk 6]
```diff
@@ -92,9 +92,9 @@ def convert_to_text(self) -> str:
     def _add_symlinks(self, destination_pig_file):
         destination_pig_file.write("set mapred.create.symlink yes;\n")
         if self.files:
-            destination_pig_file.write("set mapred.cache.file {};\n".format(self.hdfs_files))
+            destination_pig_file.write("set mapred.cache.file {};\n".format(",".join(self.hdfs_files)))
```
[the remark this answers is not in the document]
Comment (mik-laj) [reply oozie-to-airflow/mappers/pig_mapper.py]: In a separate PR I will have no reason to introduce this change.  This change has a logical justification in the context of this PR, because elsewhere the code works properly.

File: oozie-to-airflow/mappers/spark_mapper.py (hunk 1 above)
Comment (mik-laj) [reply oozie-to-airflow/mappers/spark_mapper.py]: Can you provide a link to which documentation you are looking at?

File: oozie-to-airflow/mappers/spark_mapper.py (hunk 4 above)
Comment (mik-laj) [reply oozie-to-airflow/mappers/spark_mapper.py]: I tried to avoid changes in other mappers when it is not necessary. Other mappers do not use the Task object and need to be updated to be compatible with new ideas. The change you suggest is ready. Look at: https://github.com/GoogleCloudPlatform/cloud-composer/pull/127

File: oozie-to-airflow/templates/action.tpl (hunk 2 above)
Comment (mik-laj) [reply oozie-to-airflow/templates/action.tpl]: A separate PR will have no reason to introduce this change. We agreed that every PR should be justified.

File: oozie-to-airflow/templates/mapreduce.tpl (hunk 3 above)
Comment (mik-laj) [reply oozie-to-airflow/templates/mapreduce.tpl]: A separate PR will have no reason to introduce this change. We agreed that every PR should be justified.

File: oozie-to-airflow/utils/file_archive_extractors.py [hunk 7]
```diff
@@ -57,8 +57,8 @@ class FileExtractor:
     """ Extracts all file paths from an Oozie node """
 
     def __init__(self, oozie_node: Element, params: Dict[str, str]):
-        self.files: str = ""
-        self.hdfs_files: str = ""
+        self.files: List[str] = []
+        self.hdfs_files: List[str] = []
```
[the remark this answers is not in the document]
Comment (mik-laj) [reply oozie-to-airflow/utils/file_archive_extractors.py]: We agreed that every PR should be justified. Introducing this change as a separate PR will not be justified. We agreed that commits we use as a history of changes and are not used to divide big changes into smaller ones.

File: oozie-to-airflow/templates/action.tpl (hunk 2 above)
Comment (mik-laj) [reply oozie-to-airflow/templates/action.tpl]: I introduce unification in this PR, so that there is no duplicate code in the project.

File: oozie-to-airflow/mappers/fs_mapper.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply oozie-to-airflow/mappers/fs_mapper.py]: Ok, thanks for the explanation.

File: oozie-to-airflow/utils/file_archive_extractors.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply oozie-to-airflow/utils/file_archive_extractors.py]: > We agreed that commits we use as a history of changes and are not used to divide big changes into smaller ones.
I don't recall this discussion. I think the division of PRs into commits should serve first the reviewers, to make code reviews easier, and secondly should make the history of changes clear.
If someone looks at the history of these lines and sees "Add Spark Mapper" it will not be a good explanation of what happened here.

But most importantly: **what about other mappers which rely in these variables being strings**?

File: oozie-to-airflow/mappers/pig_mapper.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply oozie-to-airflow/mappers/pig_mapper.py]: This change touches the Pig Mapper and from what I see it doesn't affect Spark Mapper in any way. Am I missing something?
If I'm correct, then I don't see any reason why it should be in the "Add Spark Mapper" PR instead of a "Fix Pig Mapper" PR.

File: oozie-to-airflow/mappers/spark_mapper.py (hunk 1 above)
Comment (sprzedwojski) [reply oozie-to-airflow/mappers/spark_mapper.py]: Sorry, my bad: https://oozie.apache.org/docs/5.1.0/DG_SparkActionExtension.html

image

File: oozie-to-airflow/templates/action.tpl (hunk 2 above)
Comment (sprzedwojski) [reply oozie-to-airflow/templates/action.tpl]: Ok, let's do this here. However for the future I'd suggest a separate commit explaining the unification of `fs.tpl` --> `action.tpl`.

File: oozie-to-airflow/templates/mapreduce.tpl (hunk 3 above)
Comment (sprzedwojski) [reply oozie-to-airflow/templates/mapreduce.tpl]: Please explain what is the reason to change the mapreduce template in the "Add Spark Mapper" PR?

File: oozie-to-airflow/mappers/spark_mapper.py (hunk 4 above)
Comment (sprzedwojski) [reply oozie-to-airflow/mappers/spark_mapper.py]: Ok, thanks for the explanation.

File: oozie-to-airflow/utils/file_archive_extractors.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mik-laj) [reply oozie-to-airflow/utils/file_archive_extractors.py]: Other mappers had to be updated to be compatible with the new signature of the method.

File: oozie-to-airflow/mappers/spark_mapper.py (hunk 1 above)
Comment (sprzedwojski) [reply oozie-to-airflow/mappers/spark_mapper.py]: We can skip this for now. Maybe let's just mention in the readme that we don't support this yet.

File: oozie-to-airflow/examples/spark/job.properties [hunk 8]
```diff
@@ -36,4 +36,4 @@ master=local[*]
 queueName=default
 examplesRoot=examples
 oozie.use.system.libpath=true
-oozie.wf.application.path=${nameNode}/user/${user.name}/${examplesRoot}/apps/spark
+oozie.wf.application.path=${nameNode}/user/${user.name}/spark
```
Comment (potiuk) [inline oozie-to-airflow/examples/spark/job.properties]: This path is not good for automated execution. It should follow the same scheme as all the others examples.

File: oozie-to-airflow/examples/spark/workflow.xml [hunk 9]
```diff
@@ -39,14 +39,14 @@
             <resource-manager>${resourceManager}</resource-manager>
             <name-node>${nameNode}</name-node>
             <prepare>
-                <delete path="${nameNode}/user/${user.name}/${examplesRoot}/output-data/spark"/>
+                <delete path="${nameNode}/user/spark/examples/spark/oozie-examples-4.3.0-copy.jar"/>
```
Comment (potiuk) [inline oozie-to-airflow/examples/spark/workflow.xml]: Why do we have spark examples here? the example should be self-sufficient and should not require the .jar file in specific directory. Can we add a file to copy to "assets" folder and copy it from there ? This is much better because we can execute the test automatically

File: oozie-to-airflow/mappers/spark_mapper.py (hunk 1 above)
Comment (potiuk) [reply oozie-to-airflow/mappers/spark_mapper.py]: Yeah. We mentioned it previously in the design as far as I remember.

File: oozie-to-airflow/mappers/spark_mapper.py [hunk 10]
```diff
@@ -45,190 +59,171 @@ def __init__(
...
             if spark_opt[0] == "conf":
                 # Splits key1=value1 into [key1, value1]
-                conf_val = spark_opt[1].split("=", maxsplit=1)
-                self.conf[conf_val[0]] = conf_val[1]
-            else:
-                self.__dict__[spark_opt[0]] = "'" + " ".join(spark_opt[1:]) + "'"
+                conf_pairs = spark_opt[1].split("=", 1)
+                # Value is required
+                if len(conf_pairs) != 2:
+                    continue
```
Comment (potiuk) [inline oozie-to-airflow/mappers/spark_mapper.py]: Should we fail here? I think this is one of the semantic checks that we should not silently ignore.

**Comment by potiuk** [comment]
Nicely split now. It's really easy to review it now when it is split to separate commits.

File: oozie-to-airflow/examples/spark/workflow.xml (hunk 9 above)
Comment (mik-laj) [reply oozie-to-airflow/examples/spark/workflow.xml]: This is a result artifact. I have to specify where I will store files.  This line provides idempotence. I will look at how to now ensure the availability of this library using your solution.

File: oozie-to-airflow/mappers/spark_mapper.py (hunk 1 above)
Comment (mik-laj) [reply oozie-to-airflow/mappers/spark_mapper.py]: I added a note in README.md

File: oozie-to-airflow/examples/spark/job.properties (hunk 8 above)
Comment (mik-laj) [reply oozie-to-airflow/examples/spark/job.properties]: Thanks. I updated a PR.

**Comment by sprzedwojski** [comment]
It's nicely split into commits, therefore I would merge it as many commits.
However the fixup commits should be manually squashed before.

File: oozie-to-airflow/README.md [hunk 11]
```diff
@@ -409,6 +409,41 @@ using the configuration elements.
...
+`configuration.properties` and fill in with configuration data.
+
+##### Output
+In this example the output will appear in `/output/spark/spark.py`.
+
+The converted DAG uses the `DataProcSparkOperator` in Airflow.
+
+#### Current limitations
+
+**1. Tasks written in Java are only supported**
```
Comment (sprzedwojski) [inline oozie-to-airflow/README.md]:
```suggestion
**1. Ony tasks written in Java are supported**
```

File: oozie-to-airflow/README.md [hunk 12]
```diff
@@ -409,6 +409,41 @@ using the configuration elements.
...
+
+The solution was tested with only a single Jar file.
+
+**2. No Spark launcher configuration**
+
+From the [Oozie documentation](https://oozie.apache.org/docs/5.1.0/DG_SparkActionExtension.html):
+> Shell launcher configuration can be specified with a file, using the job-xml element, and inline,
+using the configuration elements.
+
+Currently there is no way specify the Spark launcher configuration (it is ignored).
```
Comment (sprzedwojski) [inline oozie-to-airflow/README.md]:
```suggestion
Currently there is no way to specify the Spark launcher configuration (it is ignored).
```

File: oozie-to-airflow/README.md [hunk 13]
```diff
@@ -409,6 +409,41 @@ using the configuration elements.
...
+
+**3. No all elements is supported**
```
Comment (sprzedwojski) [inline oozie-to-airflow/README.md]:
```suggestion
**3. Not all elements are supported**
```

File: oozie-to-airflow/README.md (hunk 11 above)
Comment (mik-laj) [reply oozie-to-airflow/README.md]: I updated a PR.

File: oozie-to-airflow/README.md (hunk 12 above)
Comment (mik-laj) [reply oozie-to-airflow/README.md]: I updated a PR.

File: oozie-to-airflow/README.md (hunk 13 above)
Comment (mik-laj) [reply oozie-to-airflow/README.md]: I updated a PR. Thanks.

File: oozie-to-airflow/mappers/spark_mapper.py (hunk 10 above)
Comment (mik-laj) [reply oozie-to-airflow/mappers/spark_mapper.py]: I updated a PR. Thanks.

File: oozie-to-airflow/examples/spark/workflow.xml (hunk 9 above)
Comment (mik-laj) [reply oozie-to-airflow/examples/spark/workflow.xml]: I updated a PR. Thanks.
